### PR TITLE
trim build dependencies from output

### LINKED ISSUE #2773: Dependencies declared as type:build in project.json shouldn't end up in published output
## Steps to reproduce
Create a new ASP.NET Core web site application using current VS tooling build. Publish the application using `dotnet publish`.

## Expected  behavior
Dependencies declared as `"type": "build"` and their dependencies should **not** be included in the published output.

## Actual behavior
The assets for the dependencies declared as `"type": "build"` and their dependencies are included in the published output.
This is currently accounting for almost 5MB (~15%)in the published application output.

## Environment data
`dotnet --info` output:
```
.NET Command Line Tools (1.0.0-rc2-002426)

Product Information:
 Version:     1.0.0-rc2-002426
 Commit Sha:  08ee6dedd9

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.10586
 OS Platform: Windows
 RID:         win10-x64
```

### PR DESCRIPTION
**Pull request by anurse**
fixes #2773

/cc @pakrym

### CODE REVIEW

File: scripts/dotnet-cli-build/Utils/Crossgen.cs [hunk 1]
```diff
@@ -109,8 +109,8 @@ public void CrossgenDirectory(BuildTargetContext c, string pathToAssemblies)
 
                 ExecSilent(_crossGenPath, crossgenArgs, env);
 
-                File.Delete(file);
-                File.Move(tempPathName, file);
+                File.Copy(tempPathName, file, overwrite: true);
```
Comment (pakrym) [inline scripts/dotnet-cli-build/Utils/Crossgen.cs]: Hm?

File: src/Microsoft.DotNet.Compiler.Common/Executable.cs [hunk 2]
```diff
@@ -150,7 +149,17 @@ private void WriteDepsFileAndCopyProjectDependencies(LibraryExporter exporter)
             CopyAssets(packageExports);
         }
 
-        private void WriteRuntimeConfig(LibraryExporter exporter)
+        public void WriteConfigurationFiles(IEnumerable<LibraryExport> allExports, IList<LibraryExport> depsExports)
```
Comment (pakrym) [inline src/Microsoft.DotNet.Compiler.Common/Executable.cs]: `IEnumerable` vs `IList`?

Comment (pakrym) [inline src/Microsoft.DotNet.Compiler.Common/Executable.cs]: Lets use single collection interface in this file.

File: test/dotnet-publish.Tests/PublishAppsWithToolsTests.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+            var publishDir = publishCommand.GetOutputDirectory(portable: true);
+
+            publishDir.Should().HaveFiles(new[]
+            {
+                // This one is directly referenced
+                "Microsoft.AspNetCore.Mvc.Core.dll"
+            });
+
+            // But these are brought in only by the type:build dependency, and should not be published
```
Comment (pakrym) [inline test/dotnet-publish.Tests/PublishAppsWithToolsTests.cs]: What about package that is dependency of 'type:build' but is brought by another package as dependency too?

File: src/Microsoft.DotNet.ProjectModel/ProjectContextBuilder.cs [hunk 4]
```diff
@@ -168,14 +168,14 @@ public IEnumerable<ProjectContext> BuildAllTargets()
                 var deduper = new HashSet<string>();
                 foreach (var target in LockFile.Targets)
                 {
-                    var id = $"{target.TargetFramework}/{target.RuntimeIdentifier}";
+                    var context = Clone()
```
Comment (pakrym) [inline src/Microsoft.DotNet.ProjectModel/ProjectContextBuilder.cs]: Test for this

File: src/Microsoft.DotNet.Compiler.Common/Executable.cs [hunk 5]
```diff
@@ -37,11 +37,13 @@ public class Executable
         private readonly CommonCompilerOptions _compilerOptions;
 
         public Executable(ProjectContext context, OutputPaths outputPaths, LibraryExporter exporter, string configuration)
+            : this(context, outputPaths, outputPaths.RuntimeOutputPath, outputPaths.IntermediateOutputDirectoryPath, exporter, configuration) { }
+        public Executable(ProjectContext context, OutputPaths outputPaths, string runtimeOutputPath, string intermediateOutputDirectoryPath, LibraryExporter exporter, string configuration)
```
Comment (pakrym) [inline src/Microsoft.DotNet.Compiler.Common/Executable.cs]: Add empty line

**Comment by pakrym** [comment]
:speech_balloon: :arrow_right:  :shipit:

File: scripts/dotnet-cli-build/Utils/Crossgen.cs (hunk 1 above)
Comment (anurse) [inline scripts/dotnet-cli-build/Utils/Crossgen.cs]: It seems to be more reliable. Sometimes the delete was failing...

File: src/Microsoft.DotNet.ProjectModel/ProjectContextBuilder.cs (hunk 4 above)
Comment (anurse) [inline src/Microsoft.DotNet.ProjectModel/ProjectContextBuilder.cs]: I'm going to revert it. I don't need it in the end.

Comment (anurse) [inline src/Microsoft.DotNet.ProjectModel/ProjectContextBuilder.cs]: Never mind, added a test :)

**Pull request merged**